src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -175,225 +138,128 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
             switch (operand.Kind())
             {
-                // We have something like:
-                //
-                //      var x = e as T;
-                //      while (b != null) { ... }
-                //
-                // It's not necessarily safe to convert this to:
-                //
-                //      while (x is T b) { ... }
-                //
-                // That's because in this case, unlike the original code, we're type-checking in every iteration
-                // so we do not replace simple null check with the "is" operator if it's in a while loop
-                case SyntaxKind.IdentifierName when targetStatement.Kind() != SyntaxKind.WhileStatement:
-                {
-                    // var x = e as T;
-                    // if (x != null) F(x);
-                    var identifier = (IdentifierNameSyntax)operand;
-                    var declarator = TryFindVariableDeclarator(identifier, targetStatement, parentBlock);
-                    var initializerValue = declarator?.Initializer?.Value;
-                    if (!initializerValue.IsKind(SyntaxKind.AsExpression))
+                case SyntaxKind.IdentifierName:
                     {
-                        break;
-                    }
+                        // var x = e as T;
+                        // if (x != null) F(x);
+                        var identifier = (IdentifierNameSyntax)operand;
+                        if (!TryFindVariableDeclarator(semanticModel, identifier, out localSymbol, out declarator))
+                        {
+                            break;
+                        }
+
+                        var initializerValue = declarator.Initializer?.Value;
+                        if (!initializerValue.IsKind(SyntaxKind.AsExpression, out asExpression))
+                        {
+                            break;
+                        }
 
-                    variableDeclarator = declarator;
-                    asExpression = initializerValue;
-                    return true;
-                }
+                        return true;
+                    }
 
                 case SyntaxKind.SimpleAssignmentExpression:
-                {
-                    // T x;
-                    // if ((x = e as T) != null) F(x);
-                    var assignment = (AssignmentExpressionSyntax)operand;
-                    if (!assignment.Right.IsKind(SyntaxKind.AsExpression) ||
-                        !assignment.Left.IsKind(SyntaxKind.IdentifierName))
                     {
-                        break;
-                    }
+                        // T x;
+                        // if ((x = e as T) != null) F(x);
+                        var assignment = (AssignmentExpressionSyntax)operand;
+                        if (!assignment.Right.IsKind(SyntaxKind.AsExpression, out asExpression) ||
+                            !assignment.Left.IsKind(SyntaxKind.IdentifierName, out IdentifierNameSyntax identifier))
```

Comment:
❔ Any reason to not apply `WalkDownParentheses` here?

```csharp
if ((x = (e as T)) != null) F(x);
```

Again fine with leaving this as a follow-up item unless there is a known problem supporting that.

Reply:
That should be rare, because there's already a pair of parentheses, so it would've not help with "readability", if you will. I'll keep these in mind for the next possible bug-fix PR.